### PR TITLE
[FLINK-18844][json][maxwell] Support maxwell-json format to read Maxwell changelogs

### PR DESCRIPTION
**Pull request by lonelyGhostisdog**
## What is the purpose of the change

*support maxwell-json for Maxwell binlog collector*

## Brief change log

  - *add DeserializationSchema for deserialize byte[]*
  - *add SerializationSchema for serialize data to byte[]*
  - *add formatfactory for DeserializationSchema&SerializationSchema*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added java tests for deserialize byte[] from table source & serialize data to byte[] into table sink*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): ( no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: ( no)
  - The serializers: ( no )
  - The runtime per-record code paths (performance sensitive): ( no )
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: ( no )
  - The S3 file system connector: ( no)

## Documentation

  - Does this pull request introduce a new feature? (yes )
  - If yes, how is the feature documented? ( docs / JavaDocs )

### CODE REVIEW

**Comment by lonelyGhostisdog** [comment]
@flinkbot run azure

**Comment by lonelyGhostisdog** [comment]
@flinkbot run travis

File: flink-formats/flink-json/src/main/java/org/apache/flink/formats/json/maxwell/MaxwellJsonDeserializationSchema.java [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+
+	@Override
+	public void deserialize(byte[] message, Collector<RowData> out) throws IOException {
+		try {
+			RowData row = jsonDeserializer.deserialize(message);
+			String type = row.getString(2).toString(); // "type" field
+			if (OP_INSERT.equals(type)) {
+				// difference with canal-json is that the "data" section in canal-json contains Array <Row <>>,
+				// and maxwell-json is Row <>
+				// So just take out Row<> in row and collect it
```
Comment (wuchong) [inline flink-formats/flink-json/src/main/java/org/apache/flink/formats/json/maxwell/MaxwellJsonDeserializationSchema.java:100]: We don't need to compare with canal-json in the comment. From my point of view, canal-json is a special case in encoding. You can just mention the "data" is the new image of insertion.

File: flink-formats/flink-json/src/main/java/org/apache/flink/formats/json/maxwell/MaxwellJsonDeserializationSchema.java [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+	/** Flag indicating whether to ignore invalid fields/rows (default: throw an exception). */
+	private final boolean ignoreParseErrors;
+
+	/** Number of fields. */
+	private final int fieldCount;
+
+	public MaxwellJsonDeserializationSchema(RowType rowType,
+											TypeInformation<RowData> resultTypeInfo,
+											boolean ignoreParseErrors,
+											TimestampFormat timestampFormatOption) {
```
Comment (wuchong) [inline flink-formats/flink-json/src/main/java/org/apache/flink/formats/json/maxwell/MaxwellJsonDeserializationSchema.java:73]:
```suggestion
	public MaxwellJsonDeserializationSchema(
			RowType rowType,
			TypeInformation<RowData> resultTypeInfo,
			boolean ignoreParseErrors,
			TimestampFormat timestampFormatOption) {
```

Indent.

File: flink-formats/flink-json/src/main/java/org/apache/flink/formats/json/maxwell/MaxwellJsonSerializationSchema.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+	@Override
+	public int hashCode() {
+		return Objects.hash(jsonSerializer, timestampFormat);
+	}
+
+	private RowType createJsonRowType(DataType databaseSchema) {
+		DataType payload = DataTypes.ROW(
+			DataTypes.FIELD("data", databaseSchema),
+			DataTypes.FIELD("old", databaseSchema),
```
Comment (wuchong) [inline flink-formats/flink-json/src/main/java/org/apache/flink/formats/json/maxwell/MaxwellJsonSerializationSchema.java:108]: From the `maxwell-data.txt`, it seems that insert and delete doesn't produce `old` field. Shall we remove this?

**Comment by wuchong** [comment]
Besides, it would be great if you can add the corresponding documentation in this PR. 
But create a following up JIRA issue is also fine to me.

**Comment by lonelyGhostisdog** [comment]
> Besides, it would be great if you can add the corresponding documentation in this PR.
> But create a following up JIRA issue is also fine to me.

i will finish these bugs soon.
 Documentation will be add next PR .

**Comment by lonelyGhostisdog** [comment]
@flinkbot run azure

**Comment by lonelyGhostisdog** [comment]
> ## CI report:
> * [d2a7bde](https://github.com/apache/flink/commit/d2a7bde48d5c4e4106e9f889e0a3729c5a71c716) Azure: [FAILURE](https://dev.azure.com/apache-flink/98463496-1af2-4620-8eab-a2ecc1a2e6fe/_build/results?buildId=5379)
> 
> Bot commands

**Comment by lonelyGhostisdog** [comment]
> ## CI report:
> * [d2a7bde](https://github.com/apache/flink/commit/d2a7bde48d5c4e4106e9f889e0a3729c5a71c716) Azure: [FAILURE](https://dev.azure.com/apache-flink/98463496-1af2-4620-8eab-a2ecc1a2e6fe/_build/results?buildId=5379)
> 
> Bot commands

@flinkbot  run azure

**Comment by wuchong** [comment]
Build is passed. 
Merging...

**Pull request merged**